src/Compilers/Core/Portable/SymbolDisplay/AbstractSymbolDisplayVisitor.cs

```diff
@@ -75,8 +66,60 @@ protected AbstractSymbolDisplayVisitor NotFirstVisitorNamespaceOrType
             }
         }
 
+        protected void Initialize(
+            ArrayBuilder<SymbolDisplayPart> builder,
+            SymbolDisplayFormat format,
+            bool isFirstSymbolVisited,
+            SemanticModel? semanticModelOpt,
+            int positionOpt,
+            bool inNamespaceOrType)
+        {
+            Debug.Assert(format != null);
+
+            if (_format is not null)
+            {
+                // Should not be re-initializing a visitor which is already in use.
+                throw new InvalidOperationException();
+            }
+
+            _builder = builder;
+            _format = format;
+            _isFirstSymbolVisited = isFirstSymbolVisited;
+
+            _semanticModelOpt = semanticModelOpt;
+            _positionOpt = positionOpt;
+            _inNamespaceOrType = inNamespaceOrType;
+
+            // If we're not the first symbol visitor, then we will just recurse into ourselves.
+            if (!isFirstSymbolVisited)
+            {
+                _lazyNotFirstVisitor = this;
+            }
+        }
+
+        public virtual void Free()
+        {
+            if (_lazyNotFirstVisitor != this && _lazyNotFirstVisitor is not null)
+                FreeNotFirstVisitor(_lazyNotFirstVisitor);
+
+            if (_lazyNotFirstVisitorNamespaceOrType != this && _lazyNotFirstVisitorNamespaceOrType is not null)
+                FreeNotFirstVisitor(_lazyNotFirstVisitorNamespaceOrType);
+
+            _builder = null!;
+            _format = null!;
```

Comment:
If `_format` is the field we're using as a sentinel of "this thing is currently in use", we should set it to `null` last.

Reply:
➡️ I decided to keep these in the order the fields are declared in the type, so it's easy to see they are all included. The null check is not intended to suggest that the type is thread-safe; it merely serves to document a precondition.

Reply:
> The null check is not intended to suggest that the type is thread-safe; it merely serves to document a precondition.

It doesn't suggest the type is threadsafe, but it does suggest that we're taking some amount of care to ensure that `SymbolDisplayVisitor` isn't being reused by multiple things at the same time.

Reply:
➡️ Relocated the assignment so this can get merged